build/ArtifactBuilder/Artifacts/ZipArchive.cs

```diff
@@ -56,73 +55,73 @@ protected override void InternalBuild()
+            {
+                throw new PackagingException($"The unpacked ZIP archive was missing the expected component {missingComponent}");
+            }
 
-            // core agent
+            var unexpectedUnpackedComponents = new SortedSet<string>(unpackedComponents, StringComparer.OrdinalIgnoreCase);
+            unexpectedUnpackedComponents.ExceptWith(expectedComponents);
+            foreach (var unexpectedComponent in unexpectedUnpackedComponents)
+            {
+                throw new PackagingException($"The unpacked ZIP archive contained an unexpected component {unexpectedComponent}");
+            }
 
-            var coreSourceHomeBuilderPath = $@"{_coreAgentComponents.HomeRootPath}\newrelichome_{_coreAgentComponents.Platform}_coreclr";
+            // cleanup
+            FileHelpers.DeleteDirectories(unpackedLocation);
+        }
 
-            var rootCoreInstallDirComponents = RemoveSourceHomeBuilderPath(_coreAgentComponents.RootInstallDirectoryComponents, coreSourceHomeBuilderPath);
-            var coreAgentHomeDirComponents = RemoveSourceHomeBuilderPath(_coreAgentComponents.AgentHomeDirComponents, coreSourceHomeBuilderPath);
-            var coreExtensionDirComponents = RemoveSourceHomeBuilderPath(_coreAgentComponents.ExtensionDirectoryComponents, coreSourceHomeBuilderPath);
-            var coreWrapperXmlFiles = RemoveSourceHomeBuilderPath(_coreAgentComponents.WrapperXmlFiles, coreSourceHomeBuilderPath);
+        private SortedSet<string> GetExpectedComponents(string installedFilesRoot)
+        {
+            var expectedComponents = new SortedSet<string>(StringComparer.OrdinalIgnoreCase);
 
-            VerifyComponentsExist(rootCoreInstallDirComponents, unpackDir, null);
-            VerifyComponentsExist(coreAgentHomeDirComponents, unpackDir, CoreSubDirectoryName);
-            VerifyComponentsExist(coreExtensionDirComponents, unpackDir, CoreSubDirectoryName);
-            VerifyComponentsExist(coreWrapperXmlFiles, unpackDir, CoreSubDirectoryName);
+            // framework agent
 
-            // Open questions
-            // 1. Why does the Agent API DLL get placed in the home dir for the core agent, but not the framework agent?
-            // 2. Why do the RootInstallDirectoryComponents differ between framework and core?
-            // 3. How to account for "agentInfo.json" in the home dir (both framework and core)?
+            ValidationHelpers.AddFilesToCollectionWithNewPath(expectedComponents, installedFilesRoot, _frameworkAgentComponents.RootInstallDirectoryComponents);
 
-            // cleanup
-            FileHelpers.DeleteDirectories(unpackDir);
-        }
+            var netframeworkFolder = Path.Join(installedFilesRoot, FrameworkSubDirectoryName);
+            ValidationHelpers.AddFilesToCollectionWithNewPath(expectedComponents, netframeworkFolder, _frameworkAgentComponents.AgentHomeDirComponents);
+            expectedComponents.Add(Path.Join(netframeworkFolder, AgentInfo.AgentInfoFilename));
```

Comment:
You might not need to special case the agentinfo filename. The file is already available as a property on the frameworkcomponents class. Alternatively, the agentinfo json file is part of the ConfigurationComponents property too.

Reply:
Thanks, that works better.